File: 0133-clone-graph/0133-clone-graph.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution {
public:
    //As the graph is undirected that means if 1-> 2 edge exists, 2->1 edge will also exist.
    //And if simply try to do it recursivly we might be stuck in an infinite loop. 
    //So we create this map to see if we have already made a deep copy a node or note
    unordered_map<Node*,Node*> visited;
    Node* cloneGraph(Node* node) {
        //if node is null return
        if(node == nullptr)
            return nullptr;
        //if we already created deep copy of this node return
        if(visited.find(node) != visited.end()){
            return visited[node];
        }
        //creat deep copy
        Node* res = new Node();
        res->val = node->val;
        res->neighbors.resize(node->neighbors.size());
        //store in map
        visited[node] = res;
        //create deep copy of each node by calling hte cloneGraph function recursively.
        for(int i=0;i<node->neighbors.size();i++){
            res->neighbors[i] = cloneGraph(node->neighbors[i]);
        }
        //return the result;
        return res;
    }
};

File: 0133-clone-graph/0133-clone-graph.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

from typing import Optional
class Solution:
    def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:
        new_nodes_map = {}

        def cloneNode(current_node):
            if current_node == None:
                return None
                
            if current_node.val in new_nodes_map:
                return new_nodes_map[current_node.val]

            new_node = Node()
            new_node.val = current_node.val
            new_nodes_map[new_node.val] = new_node

            for neighbor in current_node.neighbors:
                new_node.neighbors.append(cloneNode(neighbor))
            return new_node
        return cloneNode(node)

        